### DISCUSSION TITLE
Build once run in multiple environments

### DISCUSSION THREAD

**Discussion by Tasteful** [vercel/next.js#55319] [Help]
### Summary

We want to build the next.js application into standalone output once and then run the output in different environments with different domain names.

The drawback that I have found is that the `nextConfig` is created during build time and when running the `node server.js` the static one is used so even if we building and using environment variables, example to define the `remotePatterns`, with different domain names depend on the environment variables if set during build time.

Is there a way to extend the configuration dynamic when running as standalone output, example replace the `remotePatterns` during runtime?

**Comment by leerob** [comment]
With the App Router, you can configure your root layout to run dynamically, allowing you to access runtime values from environment variables on the server.

```tsx
// app/layout.tsx

export const dynamic = 'force-dynamic'

export default function RootLayout({
  children,
}: {
  children: React.ReactNode
}) {
  const value = process.env.MY_VALUE
  return (
    <html lang="en">
      <body>{children}</body>
    </html>
  )
}
```

**Reply by devevignesh** [reply]
@leerob The approach is working fine in the server components. However, when I try to implement it in the client component like this `<Link href={process.env.APP_URL || "test-link"}>test link</Link>` the href value displays "test-link" instead of the `APP_URL`. Should I prefix the runtime values with NEXT_PUBLIC_? Could you please explain the correct approach to access the runtime values from environment variables on the client/browser?

**Reply by leerob** [reply]
The value needs to be read dynamically in a server component. From there, you can forward / cache it for usage in client components as well 👍 But the `NEXT_PUBLIC_` approach won't work because that inlines the values during the build. Does that make sense?

**Reply by devevignesh** [reply]
Yes it makes sense. Thank you

**Reply by aryzle** [reply]
Hey I'm running into the same problem here, @leerob respectfully the current answer seems off the mark. Is there a way to set `remotePatterns` in `nextConfig` dynamically at runtime? Ultimately trying to set diff `remotePatterns` per env, e.g. test, dev, staging, prod

@Tasteful 's original post explains it well

**Reply by leerob** [reply]
Any reason not to put the multiple env configurations into the single `remotePatterns` config? It accepts an array. You are correct that what I included in this example is about reading environment variables, but `remotePatterns` is different.

**Reply by aryzle** [reply]
That could work, and what my team ended up doing was passing the env var needed per env at build time (we're using Docker so passed it as a build ARG), but this would not work if the url is dependent on runtime env vars
